### PR TITLE
Updated code to use functional options

### PR DESCRIPTION
**Pull request by razamiDev**
Updated code to use functional options, added two new functions to change subscriber retention period and visibility timeout. Wrote tests for new methods.

### CODE REVIEW

**Comment by dmacthedestroyer** [comment]
Looks like unit tests failed. Please run the tests and fix

File: aws/aws_test.go [hunk 1]
```diff
@@ -175,6 +177,92 @@ func TestEnsureQueuePolicy(t *testing.T) {
...
+	testCases := []struct {
+		Duration                time.Duration
+		CallsAWS                bool
+		ExpectedErr             error
+		ExpectedRetentionPeriod int64
+	}{
+		{Duration: 5 * time.Minute, CallsAWS: true, ExpectedRetentionPeriod: 300},
+		{Duration: -1 * time.Second, ExpectedErr: ErrMessageRetentionPeriodOutOfRange},
+		{Duration: 1000 * time.Hour, ExpectedErr: ErrMessageRetentionPeriodOutOfRange},
+		{Duration: 60 * time.Second, ExpectedErr: errors.New("expected SetQueueAttributes error"), CallsAWS: true},
```
Comment (dmacthedestroyer) [inline aws/aws_test.go]: it would be useful to add test cases that exercise the boundary conditions -- in this case that would be 60 and 1209600

File: aws/aws_test.go [hunk 2]
```diff
@@ -175,6 +177,92 @@ func TestEnsureQueuePolicy(t *testing.T) {
...
+	testCases := []struct {
+		Duration                  time.Duration
+		CallsAWS                  bool
+		ExpectedErr               error
+		ExpectedVisibilityTimeout int64
+	}{
+		{Duration: 5 * time.Minute, CallsAWS: true, ExpectedVisibilityTimeout: 300},
+		{Duration: -1 * time.Second, ExpectedErr: ErrVisibilityTimeoutOutOfRange},
+		{Duration: 1000 * time.Minute, ExpectedErr: ErrVisibilityTimeoutOutOfRange},
+		{ExpectedErr: errors.New("expected SetQueueAttributes error"), CallsAWS: true},
```
Comment (dmacthedestroyer) [inline aws/aws_test.go]: same here -- add test cases for the edge cases

File: aws/subscriber.go [hunk 3]
```diff
@@ -51,15 +51,33 @@ type awsSubscriber struct {
...
+	}
+	for _, opt := range opts {
+		opt(subscriberOptions)
+	}
 	channel := make(chan pubsub.Message)
 	errChannel := make(chan error)
 	go func() {
 		defer close(channel)
-		if fErr := s.ensureFilterPolicy(filter); fErr != nil {
+		if fErr := s.ensureFilterPolicy(subscriberOptions.Filter); fErr != nil {
```
Comment (dmacthedestroyer) [inline aws/subscriber.go]: it'd be nice to be able to do these three things in a single call instead of 3 independent calls... there's a stricter AWS rate limit on these calls so we should try to minimize them.

File: grpc/server.proto [hunk 4]
```diff
@@ -14,6 +14,11 @@ message SubscribeRequest {
   string topic = 1;
   string subscription_id = 2;
   map<string, string> filter = 3;
+  int64 retention_period = 4;
+  int64 visibility_timeout = 5;
```
Comment (dmacthedestroyer) [inline grpc/server.proto]: we might want to go with [UInt64Value types](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/u-int64-value) here, since they're optional

File: examples/hello/subscriber/main.go [hunk 5]
```diff
@@ -32,7 +34,15 @@ func main() {
...
 	}
-	c, e := s.Start(context.Background(), md)
+
+	// If trying to run basic subscriber you can start like this:
+	// c, e := s.Start(context.Background())
+
+	// If you need a subscriber with a filter, different retention period, or visibility timeout you can use
+	// the following approach:
+	retentionPeriod := 100 * time.Minute   // Length of time a message can stay in the queue
+	visibilityTimeout := 200 * time.Second // Period of time during which subsriber prevents other consumers from receiving and processing the message
```
Comment (dmacthedestroyer) [inline examples/hello/subscriber/main.go]: we don't need to define all the different options here -- just specifying the filter would be enough to convey the point. Keeping the example simple and clean is very important!

File: grpc/server.go [hunk 6]
```diff
@@ -56,7 +57,11 @@ func (s *grpcWrapper) Subscribe(req *SubscribeRequest, srv PubSub_SubscribeServe
 
 	ctx, cancel := context.WithCancel(srv.Context())
 	defer cancel()
-	c, e := subscriber.Start(ctx, req.GetFilter())
+
+	// Change int64 to time.Duration
+	retentionPeriod := time.Duration(req.GetRetentionPeriod()) * time.Second
+	visibilityTimeout := time.Duration(req.GetVisibilityTimeout()) * time.Second
```
Comment (dmacthedestroyer) [inline grpc/server.go]: what if these fields are not provided? They'd get set to 0 on the server...

File: aws/aws.go [hunk 7]
```diff
@@ -207,6 +215,40 @@ func ensureQueuePolicy(queueURL, topicArn *string, sqsClient sqsiface.SQSAPI) er
 	return policyErr
 }
 
+// changeMessageRetentionPeriod, changes the length of time in seconds for which Amazon SQS retains a message.
+// periodLength the length of message retention in seconds
+// The default if not changed is 4 days or (345600), min is 60 seconds to 1209600 seconds (14 days)
+func changeMessageRetentionPeriod(queueURL *string, newDuration time.Duration, sqsClient sqsiface.SQSAPI) error {
+	d := int(newDuration.Seconds())
+	if d < 60 || d > 1209600 {
```
[the remark this answers is not in the document]
Comment (razamiDev) [reply aws/aws.go]: 4 days or 345600 seconds is the default value which the retention period is set at. The max value is 14 days or 1209600 seconds

File: grpc/server.go (hunk 6 above)
Comment (razamiDev) [reply grpc/server.go]: No I have made it so that there are default values. If the functional options "retention period" or "visibility timeout" are not provided then it just uses the default values I have set in the Start method.

File: aws/aws.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dmacthedestroyer) [reply aws/aws.go]: oh, I see... oops :)

File: grpc/server.go (hunk 6 above)
Comment (dmacthedestroyer) [reply grpc/server.go]: except that you pass them in unconditionally... if the client didn't specify one of these fields, then the grpc server will fill it in with the default value (0), which in turn gets passed into the `Start` function here, which then gets set on line 48 in `grpc/client.go`

Run a test to verify, if you wish :)

File: aws/aws_test.go [hunk 8]
```diff
@@ -175,6 +177,92 @@ func TestEnsureQueuePolicy(t *testing.T) {
...
+			}
+			continue
+		}
+		if !tc.CallsAWS {
+			t.Errorf("expected not to call AWS, but it was called")
+		}
+		expectedTimeout := *aws.String(strconv.Itoa(int(tc.Duration.Seconds())))
+		actualVisibilityTimeout := *sqsMock.spiedSetQueueAttributesInput.Attributes[sqs.QueueAttributeNameMessageRetentionPeriod]
+		if expectedTimeout != actualVisibilityTimeout {
+			t.Errorf("expected retention timeout to be %s, but was %s", expectedTimeout, actualVisibilityTimeout)
```
[the remark this answers is not in the document]
Comment (razamiDev) [reply aws/aws_test.go]: No I meant to say retention duration. I can see that the actualVisibilityTimeout should just be actualTimeout which makes it confusing.

File: aws/aws_test.go (hunk 2 above)
Comment (razamiDev) [reply aws/aws_test.go]: Good call!

File: aws/subscriber.go (hunk 3 above)
Comment (razamiDev) [reply aws/subscriber.go]: We wouldn't be able to do all 3 calls together because the filter goes through sns not sqs, and we need the sqs call for (retention and visibility). But we can do the 2 (visibility and retention) together.

File: aws/aws.go [hunk 9]
```diff
@@ -232,10 +232,10 @@ func changeMessageRetentionPeriod(queueURL *string, newDuration time.Duration, s
 	return err
 }
 
-// changeVisibilityTimeout, Extends the message visibility for the whole queue
+// ensureVisibilityTimeout extends the message visibility for the whole queue
 // newDuration: The new value for the message's visibility timeout.
 // Value needs to be 0 to 43200 seconds.
-func changeVisibilityTimeout(queueURL *string, newDuration time.Duration, sqsClient sqsiface.SQSAPI) error {
+func ensureVisibilityTimeout(queueURL *string, newDuration time.Duration, sqsClient sqsiface.SQSAPI) error {
```
Comment (dmacthedestroyer) [inline aws/aws.go]: You could probably remove these functions now that they've been combined in the single call. Are they being used elsewhere?

File: aws/aws_test.go [hunk 10]
```diff
@@ -220,7 +222,63 @@ func TestChangeMessageRetentionPeriod(t *testing.T) {
 	}
 }
 
-func TestChangeVisibilityTimeout(t *testing.T) {
+func TestEnsureQueueAttributes(t *testing.T) {
```
Comment (dmacthedestroyer) [inline aws/aws_test.go]: great test, nice job :)

File: grpc/server.go [hunk 11]
```diff
@@ -57,11 +57,21 @@ func (s *grpcWrapper) Subscribe(req *SubscribeRequest, srv PubSub_SubscribeServe
...
+		subscriberOptions = append(subscriberOptions, pubsub.Filter(req.GetFilter()))
+	}
+
+	c, e := subscriber.Start(ctx, subscriberOptions...)
 
-	// Change int64 to time.Duration
-	retentionPeriod := time.Duration(req.GetRetentionPeriod()) * time.Second
-	visibilityTimeout := time.Duration(req.GetVisibilityTimeout()) * time.Second
-	c, e := subscriber.Start(ctx, pubsub.VisibilityTimeout(visibilityTimeout), pubsub.RetentionPeriod(retentionPeriod), pubsub.Filter(req.GetFilter()))
 	log.Printf("GRPC: starting subscription for topic %q, subID %q, filter %v", req.GetTopic(), req.GetSubscriptionId(), req.GetFilter())
```
Comment (dmacthedestroyer) [inline grpc/server.go]: request: change this log to:
``` go
log.Printf("GRPC: starting subscription: %v", req)
```

File: aws/aws.go (hunk 9 above)
Comment (razamiDev) [reply aws/aws.go]: Not at the moment. I figured in the future maybe it could be used somewhere. If we want to expose a client to change the subscriber timeout or retention specifically then we can.
